### ISSUE TITLE
[Client bug]: Release build with SDK omits the Issuer supplied in call to add user

### ISSUE THREAD

**Issue by twainj** [microsoftgraph/msgraph-sdk-dotnet#1267]
**Describe the bug**
In release builds, our calls to create a user with the GraphServiceClient are failing with a ServiceException. The service is responding with a 400 Bad Request, Message: "issuer cannnot be null or empty".
IMPORTANT: This only happens when issued from a release build of our project. 

**To Reproduce**
1. In our project (an Asp.Net 6 web app), we have an action that uses the SDK to try and create a user: `GraphClient.Users.Request().AddAsync(graphUser)`
2. The supplied graph user has the issuer set in its Identities member:
```
graphUser.Identities = new List<GraphApi.ObjectIdentity>
                {
                    new GraphApi.ObjectIdentity
                    {
                        SignInType = "emailAddress",
                        Issuer = "<tenant>.onmicrosoft.com",
                        IssuerAssignedId = contact.Email
                    }
                }
```
3. When the site is run from a Debug build, the user is created as expected.
4. When the site is run from a Release build, the user is not created, and the 400 error occurs ("issuer cannnot be null or empty")
5. Tracing the requests that are being sent confirms that the `Issuer` member is empty in the request sent from the Release build.

**Expected behavior**
We expected the same request to result from the client, regardless of the type of build.

**Client version**
First noted on version 4.15. Replicated on 4.18 and 4.19.

**Desktop (please complete the following information):**
 - OS: Windows 10 Pro
 - Browser: Chrome v98

**Comment by darrelmiller** [comment]
Do you have any thoughts as to what might be causing this?  @andrueastman   I verified the CSDL and I don't see any issues.

**Comment by andrueastman** [comment]
I wasn't able to repro this on a NET6.0 release build as the generated payloads are the same.

> Tracing the requests that are being sent confirms that the Issuer member is empty in the request sent from the Release build.

@twainj Any chance you can clarify if the payload is missing the entire "issuer" property (this would possibly point to the property being null at the time of serialization)

```json
{
	"identities": [{      //issuer propery missing
		"issuerAssignedId": "email",
		"signInType": "emailAddress"
	}],
	"@odata.type": "microsoft.graph.user"
}
```
or is the property present but with an empty string?
```json
{
	"identities": [{
		"issuer": "", // set with empty string
		"issuerAssignedId": "email",
		"signInType": "emailAddress"
	}],
	"@odata.type": "microsoft.graph.user"
}
```

**Comment by twainj** [comment]
@andrueastman , It has been long enough, that I can't say for sure. I have emailed Michael Mushat, with Azure support. He was the one that did some troubleshooting with me, and found the request that raised the error. He might remember better those specifics.

My memory is that it was either simply missing, or perhaps was explicitly set to null:

```
{
	"identities": [{
		"issuer": null, // explicitly set
		"issuerAssignedId": "email",
		"signInType": "emailAddress"
	}],
	"@odata.type": "microsoft.graph.user"
}
```

But I'll update when I hear back from him, in case he has better notes.

**Comment by twainj** [comment]
Hi @andrueastman , I was able to get some logging on the request, and I can confirm that the Issuer is indeed missing.

Here's a sample request that was sent with a Debug build:
```
{
	"displayName": "jonesy@twainj.net",
	"givenName": "Jonesy",
	"identities": [{
			"issuer": "springfinb2c.onmicrosoft.com",
			"issuerAssignedId": "jonesy@twainj.net",
			"signInType": "emailAddress"
		}
	],
	"passwordPolicies": "DisablePasswordExpiration",
	"passwordProfile": {
		"forceChangePasswordNextSignIn": false,
		"password": "..."
	},
	"surname": "Jones",
	"@odata.type": "microsoft.graph.user",
	"extension_..._requiresMigration": false
}
```

Here's the same request made with the Release build, note the missing issuer:
```
Release Build:
{
	"displayName": "jonesy@twainj.net",
	"givenName": "Jonesy",
	"identities": [{ // Issuer is missing
			"issuerAssignedId": "jonesy@twainj.net",
			"signInType": "emailAddress"
		}
	],
	"passwordPolicies": "DisablePasswordExpiration",
	"passwordProfile": {
		"forceChangePasswordNextSignIn": false,
		"password": "..."
	},
	"surname": "Jones",
	"@odata.type": "microsoft.graph.user",
	"extension_..._requiresMigration": false
}
```

**Comment by MIchaelMainer** [comment]
@twainj I'm curious if with a release build will the following emit the issuer property:

```csharp
graphUser.Identities = new List<GraphApi.ObjectIdentity>
                {
                    new GraphApi.ObjectIdentity
                    {
                        SignInType = "emailAddress",
                        //Issuer = "<tenant>.onmicrosoft.com",
                        IssuerAssignedId = contact.Email,
		        AdditionalData = new Dictionary<string, object>()
		       {
			      {"issuer","<tenant>.onmicrosoft.com" }
		       }
                    }
                }
```

**Comment by twainj** [comment]
Hi @MIchaelMainer, I gave this a shot, and it adds the issuer to the request, but not in the identities collection, where it is expected:
```
		{
			"displayName": "jonesy@twainj.net",
			"givenName": "Jonesy",
			"identities": [{
					"issuerAssignedId": "jonesy@twainj.net",
					"signInType": "emailAddress"
				}
			],
			"passwordPolicies": "DisablePasswordExpiration",
			"passwordProfile": {
				"forceChangePasswordNextSignIn": false,
				"password": "..."
			},
			"surname": "Jones",
			"@odata.type": "microsoft.graph.user",
			"issuer": "springfinb2c.onmicrosoft.com",  // Added here, instead
			"extension_..._requiresMigration": false
		}
```

**Comment by MIchaelMainer** [comment]
@twainj Is it possible that you put it in the user.AdditionalData property instead of the objectIdentityAdditionalData property? I'm curious if you will see "issuer" serialized for both debug and release profiles when you run this:

```csharp
var identity = GraphApi.ObjectIdentity
                    {
                        SignInType = "emailAddress",
                       /Issuer = "<tenant>.onmicrosoft.com",
                        IssuerAssignedId = contact.Email
                    }
	string serializedIdentity = client.HttpProvider.Serializer.SerializeObject(identity);
	// Check that serializedIdentity contains "issuer" for both debug and release profiles.
```

**Comment by twainj** [comment]
Ah, thank you, Michael, I didn't notice you were asking about putting the AdditionalData on the identity.
I tried a few things, and I think you are right, it must be the Serialization that is going wrong, as the Issuer property - in the original case - is missing in that serialization (for release builds).

When I add it to the AdditionalData (on the identity), it behaves differently. In normal Debug builds, it adds it to the request in the way you would expect. But on the Release build, the `issuer` member does appear, but now it is null:

```
	{
		"issuerAssignedId": "jonesy@twainj.net",
		"signInType": "emailAddress",
		"issuer": null
	}
```

**Comment by andrueastman** [comment]
Thanks for this @twainj. 

It looks like the serializer is behaving as expected if the property was already null as it does not emit null properties unless its added to the AdditionalData. 

Just to clarify, any chance you can add a log/debug statement and confirm the `contact.Email` is not null before it is set on the identity in the release build?

**Comment by twainj** [comment]
Ah! It was an error in our project files. To answer your question, yes, the `contact.Email` value was not null, which is why the `issuerAssignedId` was working correctly. 

But I had simplified the `issuer` member in my first post. It was actually pulling from our App Configuration settings. And it appears that a some of our conditional build symbols - which affect the config file that is bound in a very specific case that did apply here - were missing in the project settings, so the app config was being pulled incorrectly.

So, in short, the variable getting assigned to the `Issuer`, was indeed null, for release builds. As the config was being bound correctly in the rest of the app, I didn't even consider that it might not be bound correctly here. 

So I can eliminate the issue by fixing the binding in this case.

Thank you everyone for your time.

**Comment by andrueastman** [comment]
No worries @twainj.

Thanks for clearing that up.
